gateway-api-integrator/lib/charms/gateway_api_integrator/v0/gateway_route.py

```diff
@@ -0,0 +1,475 @@
+
+
+class GatewayRouteEndpointsUpdatedEvent(RelationEvent):
+    """Event representing that the proxied endpoints have been updated."""
+
+
+class GatewayRouteProviderEvents(ObjectEvents):
+    """Container for IPA Provider events."""
+
+    data_provided = EventSource(GatewayRouteAppDataProvidedEvent)
+    data_removed = EventSource(GatewayRouteAppDataRemovedEvent)
+    endpoints_updated = EventSource(GatewayRouteEndpointsUpdatedEvent)
+
+
+class IngressError(RuntimeError):
+    """Base class for custom errors raised by this library."""
+
+
+class NotReadyError(IngressError):
+    """Raised when a relation is not ready."""
+
+
+class DataValidationError(IngressError):
+    """Raised when data validation fails on IPU relation data."""
+
+
+class IngressUrl(BaseModel):
+    """Ingress url schema."""
+
+    url: AnyHttpUrl
+
+
+class GatewayRouteProviderAppData(_DatabagModel):
+    """Ingress application databag schema."""
+
+    ingress: Optional[IngressUrl] = None
+
+
+class GatewayRouteProvider(_GatewayRouteBase):
+    """Implementation of the provider of gateway-route."""
+
+    on = GatewayRouteProviderEvents()  # type: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """Constructor for GatewayRouteProvider.
+
+        Args:
+            charm: The charm that is instantiating the instance.
+            relation_name: The name of the relation endpoint to bind to
+                (defaults to "gateway-route").
+        """
+        super().__init__(charm, relation_name)
+
+    def _handle_relation(self, event: RelationEvent) -> None:
+        # created, joined or changed: if remote side has sent the required data:
+        # notify listeners.
+        if self.is_ready(event.relation):
+            data = self.get_data(event.relation)
+            self.on.data_provided.emit(  # type: ignore
+                event.relation,
+                data.app.name,
+                data.app.model,
+                data.app.hostname,
+                data.app.strip_prefix or False,
+                data.app.redirect_https or False,
+            )
+
+    def _handle_relation_broken(self, event: RelationEvent) -> None:
+        self.on.data_removed.emit(event.relation, event.relation.app)  # type: ignore
+
+    def wipe_gateway_route_data(self, relation: Relation) -> None:
+        """Clear gateway-route data from relation."""
+        assert self.unit.is_leader(), "only leaders can do this"
+        try:
+            relation.data
+        except ModelError as e:
+            logger.warning(
+                "error {} accessing relation data for {!r}. "
+                "Probably a ghost of a dead relation is still "
+                "lingering around.".format(e, relation.name)
+            )
+            return
+        del relation.data[self.app]["gateway-route"]
+        self.on.endpoints_updated.emit(relation=relation, app=relation.app)
```

Comment:
We won't need this method for the provider I think, since when we're dumping the pydantic model we have the option to specify ( clear = true ). and the provider databag is only the list of endpoints